crates/subspace-farmer/src/farmer_cache/piece_cache_state.rs

```diff
@@ -0,0 +1,150 @@
+            None => {
+                // Sort piece caches by number of stored pieces to fill those that are less
+                // populated first
+                let mut sorted_backends = self
+                    .backends
+                    .iter_mut()
+                    .enumerate()
+                    .filter_map(|(cache_index, backend)| {
+                        Some((CacheIndex::try_from(cache_index).ok()?, backend))
+                    })
+                    .collect::<Vec<_>>();
+                sorted_backends.sort_unstable_by_key(|(_, backend)| backend.free_size());
+                sorted_backends
+                    .into_iter()
+                    .rev()
+                    .find_map(|(cache_index, backend)| {
+                        backend
+                            .next_free()
+                            .map(|free_offset| FarmerCacheOffset::new(cache_index, free_offset))
+                    })
+            }
+        }
+    }
+
+    pub(super) fn get_stored_piece(
+        &self,
+        key: &RecordKey,
+    ) -> Option<&FarmerCacheOffset<CacheIndex>> {
+        self.stored_pieces.get(key)
+    }
+
+    pub(super) fn contains_stored_piece(&self, key: &RecordKey) -> bool {
+        self.stored_pieces.contains_key(key)
+    }
+
+    pub(super) fn push_stored_piece(
+        &mut self,
+        key: RecordKey,
+        cache_offset: FarmerCacheOffset<CacheIndex>,
+    ) -> Option<FarmerCacheOffset<CacheIndex>> {
+        self.stored_pieces.insert(key, cache_offset)
+    }
+
+    pub(super) fn stored_pieces(&self) -> Values<'_, RecordKey, FarmerCacheOffset<CacheIndex>> {
+        self.stored_pieces.values()
+    }
+
+    pub(super) fn remove_stored_piece(
+        &mut self,
+        key: &RecordKey,
+    ) -> Option<FarmerCacheOffset<CacheIndex>> {
+        self.stored_pieces.remove(key)
+    }
+
+    pub(super) fn free_stored_piece_if<F>(&mut self, pred: F)
+    where
+        F: FnMut(&RecordKey, &mut FarmerCacheOffset<CacheIndex>) -> bool,
+    {
+        self.stored_pieces
+            .extract_if(pred)
+            .for_each(|(_piece_index, offset)| {
+                // There is no need to adjust the `last_stored_offset` of the `backend` here,
+                // as the free_offset will be preferentially taken from the dangling free offsets
+                self.dangling_free_offsets.push_back(offset);
+            })
+    }
+
+    pub(super) fn push_dangling_free_offset(&mut self, offset: FarmerCacheOffset<CacheIndex>) {
+        trace!(?offset, "Pushing dangling free offset");
+        self.dangling_free_offsets.push_back(offset);
+    }
+
+    pub(super) fn get_backend(&self, cache_index: CacheIndex) -> Option<&CacheBackend> {
+        self.backends.get(usize::from(cache_index))
+    }
+
+    pub(super) fn backends(&self) -> impl Iterator<Item = &CacheBackend> {
+        self.backends.iter()
+    }
+
+    pub(super) fn backends_count(&self) -> usize {
+        self.backends.len()
+    }
```

Comment:
First should be `ExactSizeIterator` and second method should not exist because you can get it trivially from the iterator.